### PR TITLE
Detect Autofills in Webpages.

### LINKED ISSUE #371: Autofill credentials in web pages
## Acceptance Criteria
When the `AutofillService` encounters a `FillRequest` where the top-level `ViewNode` is a `WebView` or `GeckoView`, optimize credential filling to:
- use the domain of the displayed URL in the webpage rather than the app package name
- search for autofill-able views in a web context rather than a native Android app context

#### Documentation

- [Autofill Framework root document](https://developer.android.com/guide/topics/text/autofill)
- [Web Security Documentation](https://developer.android.com/reference/android/service/autofill/AutofillService#web-security)
- [Filling Notes](https://developer.android.com/guide/topics/text/autofill-services#fill)

#### Engineering Tasks
- [ ] Modify `ParsedStructure.kt` logic to check for `WebView`s.
- [ ] Use the `AssistStructure.ViewNode.getWebDomain()` from the `WebView` node to determine potential credentials
- [ ] Update the username and password keywords to include attributes from the wc3 standard and any encountered or relevant characteristics (more in the [autofill docs](https://developer.android.com/guide/topics/text/autofill-services#fill) )
- [ ] Update the `architecture/autofill` doc with a plaintext writeup of the heuristics that you used

#### Gotchas
There's likely a high degree of complexity here in doing the webview parsing; I'd guess more so in browsers & web-optimized pages than mobile apps that use `WebView`s for authentication like we do in Lockbox.

### PR DESCRIPTION
**Pull request by jhugman**
Fixes #371.

### CODE REVIEW

File: app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt [hunk 1]
```diff
@@ -86,35 +88,31 @@ class LockboxAutofillService(
...
-        val dataset = datasetForPossibleValues(possibleValues, parsedStructure)
-        // future parts of this method include adding any authentication steps
+        val builder = FillResponse.Builder()
 
-        return FillResponse.Builder()
-            .addDataset(dataset)
-            .build()
+        possibleValues
+            .map { serverPasswordToDataset(parsedStructure, it) }
+            .map(builder::addDataset)
```
Comment (eliserichards) [inline app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: Is it ok to use two `maps` concurrently?

Comment (jhugman) [reply app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: Yes. It looks a bit weird. Strictly, I should use a `forEach`, but passing `.forEach(builder::addDataset)` didn't compile.

File: app/src/main/java/mozilla/lockbox/extensions/ViewNode+.kt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import android.os.Build
+
+@TargetApi(Build.VERSION_CODES.O)
+fun AssistStructure.ViewNode.dump(): String {
+    val sb = StringBuilder("<?xml version=\"1.0\" encoding=\"UTF-8\"?>")
+    return dumpNode(sb).toString()
+}
+
+@TargetApi(Build.VERSION_CODES.O)
+private fun AssistStructure.ViewNode.dumpNode(sb: StringBuilder = StringBuilder()): StringBuilder {
```
Comment (eliserichards) [inline app/src/main/java/mozilla/lockbox/extensions/ViewNode+.kt]: The name `AssistStructure` confuses me. This is only for assisting autofill attributes?

File: app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt [hunk 3]
```diff
@@ -86,35 +88,31 @@ class LockboxAutofillService(
...
+        }
 
-            parsedStructure.passwordId?.let {
-                datasetBuilder.setValue(it, AutofillValue.forText(credential.password), passwordPresentation)
-            }
+        parsedStructure.passwordId?.let {
+            datasetBuilder.setValue(it, AutofillValue.forText(credential.password), passwordPresentation)
         }
 
+        log.info("presentation for $packageName: user ${credential.username}")
```
[the remark this answers is not in the document]
Comment (jhugman) [reply app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: 👍 removed logging of PII.

File: app/src/main/java/mozilla/lockbox/support/ParsedStructure.kt [hunk 4]
```diff
@@ -1,113 +1,10 @@
...
-                    return true
-                }
-            }
-        }
-
-        return false
-    }
-}
\ No newline at end of file
+data class ParsedStructure(
```
Comment (eliserichards) [inline app/src/main/java/mozilla/lockbox/support/ParsedStructure.kt]: Ditto with `ParsedStructure` confusion. Is this the "final" autofill object?

File: app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt (hunk 1 above)
Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: Makes sense :)

**Comment by jhugman** [comment]
Landing to unblock #216.

**Pull request merged**